### PR TITLE
feat(module-federation): consolidate module federation utils into module-federation package

### PR DESCRIPTION
**Pull request by Coly010**
- feat(module-federation): consolidate module federation utils into module-federation package
- chore(module-federation): fix tests and linting

## Current Behavior

Our current support for Module Federation relies on utilities that are spread and duplicated across the `@nx/webpack` package and the `@nx/rspack` package.

## Expected Behavior

Now that we have a `@nx/module-federation` package, dedupe the utils and consolidate them into a single package

## Todo
- [x] Migrations for React + Angular to install `@nx/module-federation` and point `ModuleFederationConfig` export to that package from webpack.config and rspack.config files

### CODE REVIEW

File: packages/angular/src/migrations/update-20-2-0/migrate-mf-imports-to-new-package.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+const MF_IMPORT_TO_UPDATE = 'ModuleFederationConfig';
+const MF_CONFIG_IMPORT_SELECTOR = `ImportDeclaration:has(StringLiteral[value=@nx/webpack]):has(Identifier[name=ModuleFederationConfig])`;
+const IMPORT_TOKENS_SELECTOR = `ImportClause ImportSpecifier`;
+const MF_CONFIG_IMPORT_SPECIFIER_SELECTOR = `ImportClause ImportSpecifier > Identifier[name=ModuleFederationConfig]`;
+const WEBPACK_IMPORT_SELECTOR = `ImportDeclaration > StringLiteral[value=@nx/webpack]`;
+
+export default async function migrateMfImportsToNewPackage(tree: Tree) {
+  visitNotIgnoredFiles(tree, '', (filePath) => {
+    let contents = tree.read(filePath, 'utf-8');
```
Comment (leosvelperez) [inline packages/angular/src/migrations/update-20-2-0/migrate-mf-imports-to-new-package.ts:13]: We could avoid reading all files by first checking the `filePath` extension. Additionally, we could also only visit the files of projects that have `@nx/webpack` as a dependency.

File: packages/react/src/migrations/update-20-2-0/migrate-mf-imports-to-new-package.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+const MF_IMPORT_TO_UPDATE = 'ModuleFederationConfig';
+const MF_CONFIG_IMPORT_SELECTOR = `ImportDeclaration:has(StringLiteral[value=@nx/webpack]):has(Identifier[name=ModuleFederationConfig]),ImportDeclaration:has(StringLiteral[value=@nx/rspack/module-federation]):has(Identifier[name=ModuleFederationConfig])`;
+const IMPORT_TOKENS_SELECTOR = `ImportClause ImportSpecifier`;
+const MF_CONFIG_IMPORT_SPECIFIER_SELECTOR = `ImportClause ImportSpecifier > Identifier[name=ModuleFederationConfig]`;
+const WEBPACK_IMPORT_SELECTOR = `ImportDeclaration > StringLiteral[value=@nx/webpack]`;
+const RSPACK_IMPORT_SELECTOR = `ImportDeclaration > StringLiteral[value=@nx/rspack/module-federation]`;
+
+export default async function migrateMfImportsToNewPackage(tree: Tree) {
+  visitNotIgnoredFiles(tree, '', (filePath) => {
+    let contents = tree.read(filePath, 'utf-8');
```
Comment (leosvelperez) [inline packages/react/src/migrations/update-20-2-0/migrate-mf-imports-to-new-package.ts:14]: Same suggestions as the other migration to improve performance.

**Review by leosvelperez — CHANGES_REQUESTED** [review]

File: packages/module-federation/src/utils/share.ts [hunk 3]
```diff
@@ -20,17 +20,21 @@ import {
 } from '@nx/devkit';
 import { existsSync } from 'fs';
 import type { PackageJson } from 'nx/src/utils/package-json';
+import { NormalModuleReplacementPlugin as RspackNormalModuleReplacementPlugin } from '@rspack/core';
+import { NormalModuleReplacementPlugin as WebpackNormalModuleReplacementPlugin } from 'webpack';
```
[the remark this answers is not in the document]
Comment (Coly010) [reply packages/module-federation/src/utils/share.ts:24]: Unfortunately, we can't. 

`require('@rspack/core')` results in an error where the native binary is not resolved correctly.

File: packages/angular/src/utils/mf/with-module-federation.ts [hunk 4]
```diff
@@ -62,7 +62,7 @@ export async function withModuleFederation(
               ? [
                   ...(configOverride?.runtimePlugins ?? []),
                   require.resolve(
-                    '@nx/webpack/src/utils/module-federation/plugins/runtime-library-control.plugin.js'
+                    '@nx/module-federation/src/utils/plugins/runtime-library-control.plugin.js'
```
[the remark this answers is not in the document]
Comment (Coly010) [reply packages/angular/src/utils/mf/with-module-federation.ts:65]: I'm going to vote no for now. It's internal and controlled already via a public api in `withModuleFederation`.
 If we add more, maybe, but don't see the need to pre optimize

**Review by leosvelperez — APPROVED** [review]

**Review by FrozenPandaz — APPROVED** [review]

**Review by ndcunningham — APPROVED** [review]

**Review by jaysoo — APPROVED** [review]

**Pull request merged**